src/devices/cpu/m6809/6x09dasm.cpp

```diff
@@ -57,42 +57,60 @@ const m6x09_base_disassembler::opcodeinfo *m6x09_base_disassembler::fetch_opcode
 {
 	uint16_t page = 0;
 	const opcodeinfo *op = nullptr;
+
 	while(!op)
 	{
 		// retrieve the opcode
 		uint16_t opcode = page | opcodes.r8(p++);
+		m6x09_instruction_level level = m_level;
 
 		// perform the lookup
 		auto iter = std::find_if(
 			m_opcodes.begin(),
 			m_opcodes.end(),
-			[opcode](const opcodeinfo &info) { return info.opcode() == opcode; });
+			[opcode,level](const opcodeinfo &info) { return (info.opcode() == opcode) && (info.level() <= level); });
 
 		// did we find something?
 		if (iter == m_opcodes.end())
-			return nullptr;
-
-		// was this a $10 or $11 page?
-		switch (iter->mode())
 		{
-		case PG1:
-			page = 0x1000;
-			break;
-
-		case PG2:
-			page = 0x1100;
-			break;
+			// on the 6809 an unimplemented page 2 or 3 opcodes fall thru to the first page.
+			if ( (m_level < HD6309_EXCLUSIVE) && (page != 0))
+			{
+				// backup the opcode pointer and disassemble the bare opcode.
+				--p;
+				page = 0;
+			}
+			else
+				// nothing to disassemble.
+				return nullptr;
+		}
+		else
+		{
+			// was this a $10 or $11 page?
+			switch (iter->mode())
+			{
+			case PG2:
+			case PG3:
+				// remember the page that comes first
+				if (page == 0)
+					page = iter->opcode() << 8;
+				break;
 
-		default:
-			op = &*iter;
-			break;
+			default:
+				op = &*iter;
+				break;
+			}
 		}
 	};
 
-	// is this an HD6309 exclusive instruction, and we're not HD6309?  if so, reject it
+	// is this an HD6309 exclusive instruction, and we're not HD6309? if so, reject it
 	if (op && (op->level() > m_level))
 		op = nullptr;
 
+	// is this a M6809 undocumented instruction and we're in HD6309 mode? if so, reject it
+	if (op && (op->level() == M6809_UNDOCUMENTED) && (m_level == HD6309_EXCLUSIVE))
+		op = nullptr;
```

Comment:
The greater than/less than level comparisons don’t really make that much sense any more, as they’re no longer supersets in ascending order.